draft-ietf-quic-http.md

```diff
@@ -799,25 +799,34 @@ indicated request.  This trades off network usage against a potential latency
 gain.  HTTP/3 server push is similar to what is described in HTTP/2 {{?HTTP2}},
 but uses different mechanisms.
 
-Each server push is identified by a unique Push ID. This Push ID is used in one
-or more PUSH_PROMISE frames (see {{frame-push-promise}}) that carry the request
-fields, then included with the push stream which ultimately fulfills those
-promises. When the same Push ID is promised on multiple request streams, the
-decompressed request field sections MUST contain the same fields in the
-same order, and both the name and the value in each field MUST be exact
-matches.
-
-Server push is only enabled on a connection when a client sends a MAX_PUSH_ID
-frame; see {{frame-max-push-id}}. A server cannot use server push until it
-receives a MAX_PUSH_ID frame. A client sends additional MAX_PUSH_ID frames to
-control the number of pushes that a server can promise. A server SHOULD use Push
-IDs sequentially, starting at 0. A client MUST treat receipt of a push stream
-with a Push ID that is greater than the maximum Push ID as a connection error of
-type H3_ID_ERROR.
+Each server push is identified by a unique Push ID, which is used to refer to
+the push in various contexts throughout the lifetime of the connection.
+
+The Push ID space begins at zero, and ends at a maximum value set by the
+MAX_PUSH_ID frame; see {{frame-max-push-id}}.  Server push is only enabled on a
+connection when a client sends a MAX_PUSH_ID frame.  A client sends additional
+MAX_PUSH_ID frames to control the number of pushes that a server can promise.  A
+server SHOULD use Push IDs sequentially.  A client MUST treat receipt of a push
```

Comment:
It might be beneficial to keep the "starting with 0" phrase at the end of the sentence, because the first sentence of this paragraph does not explicitly prescribe that a server SHOULD start with zero.

Reply:
I think both old and new are correct: which means there was a potential bug in the spec. Reasoning: my old reading was `server SHOULD start at 0 && use sequentially` but the new text seems slightly different. If the client and server disagree about how the ID space is used then there is the potential for interoperability problems. If a client thinks the first push ID is 0, then it can make assumptions about its accounting. If a server wants to send the first push ID as non-zero then there is no signal for it to ask for a credit window shift - it's just blocked. This seems to invalidate the statement that push is only enabled when MAX_PUSH_ID is sent. 

I'm wondering if MUST start at 0 makes sense here, even if that means a separate design issue.

Reply:
I agree that this would be an interoperability issue.  However, I think SHOULD is just fine here.  In my mind MUST is something that the peer should be able to enforce, and since streams can arrive out of order, in this case the client cannot detect if the server starts push IDs from a positive integer or if the promise with push ID 0 is delayed.

I think the statement that push is only enabled when MAX_PUSH_ID is sent is still valid.  If no MAX_PUSH_ID is sent, push is not enabled.  If a MAX_PUSH_ID is sent by the client with any value, server push is technically enabled, even if a server that insists on starting push IDs from a larger integer would be not willing to push.  Am I misunderstanding your statement?

That being said, this sentence might flow better if rephrased as "In particular, a server is not allowed to push until it receives a MAX_PUSH_ID frame.".

Reply:
Good point about detecting difficulties